src/python/WMComponent/TaskArchiver/CleanCouchPoller.py

```diff
@@ -380,6 +380,19 @@ def deleteWorkflowFromWMBSAndDisk(self):
         deletableWorkflowsDAO = self.daoFactory(classname="Workflow.GetDeletableWorkflows")
         deletablewfs = deletableWorkflowsDAO.execute()
 
+        # For T0 subtract the workflows which are not having all their blocks deleted yet:
+        if not self.useReqMgrForCompletionCheck:
+            deletedBlocksByWorkflowDAO = self.daoFactory(classname="Workflow.GetDeletedBlocksByWorkflow")
+            deletedBlocksByWorkflow = deletedBlocksByWorkflowDAO.execute()
+            for workflow in list(deletablewfs):
```

Comment:
> To avoid changing the dictionary size while iterating over it

This is not what is happening here. We are not iterating through the dictionary but rather on list of workflow names created out of the dictionary keys.

> I'd suggest to create a new plain list to hold the workflow names for workflows passing the block constraint.

This is exactly what is already implemented here. Here is exactly the place where this list is created, but rather than holding another variable in memory to keep the workflows names already passing the creteria and then reiterating on that list again to take the action, we take the action immediately at place once the criteria is satisfied.

Reply:
As you can see, it's a subtle and tricky detail ;) Lines like:
```
for workflow in list(deletablewfs):
     deletablewfs.pop(workflow)
```

use data from the dictionary (expanded in a new list object, as you pointed out) and deletes items from the dictionary.

IMO, it would be less error prone and more clear to have something like:
```
finalDeletableWfs = []
for workflow in list(deletablewfs):
     finalDeletableWfs.append(workflow)
```

it will hold not more than 100s of workflow names, so memory footprint shouldn't be even 1MB for this new object.

Reply:
Hi @amaltaro  I know this looks more readable and usually this is also my prefered way of 
There was a reason I did it by popping elements out of the already existing dictionary. 
* First the `deletableWorkflows` is not just a list of workflow names but a dictionary with workflow names used as  a key. And behind every key/workflowname sits a little bit larger structure.
* I wanted to preserve variable names and existing code structure as it is and to avoid changes in the rest of the already working and well tested code. And also keeping the code change just as 3 lines of complementary check to the list of deletable workflows makes it much more readable and easy to grasp the whole picture.

Reply:
Ok, sounds good to me. Thanks